cms/auth/utils.py

```diff
@@ -0,0 +1,114 @@
+import base64
+import json
+import logging
+from collections.abc import Iterable
+from typing import Optional
+
+import requests
+from cryptography.hazmat.backends import default_backend
+from cryptography.hazmat.primitives import serialization
+from cryptography.hazmat.primitives.asymmetric.types import PublicKeyTypes
+from django.conf import settings
+from jwt import InvalidTokenError, get_unverified_header
+from jwt import decode as jwt_decode
+from jwt.exceptions import ExpiredSignatureError
+
+from cms.core.cache import memory_cache
+
+logger = logging.getLogger(__name__)
+
+JWKS_URL = f"{settings.IDENTITY_API_BASE_URL}/jwt-keys"
+EXPECTED_ISSUER = f"https://cognito-idp.eu-west-2.amazonaws.com/{settings.AWS_COGNITO_USER_POOL_ID}"
+EXPECTED_AUDIENCE = settings.AWS_COGNITO_APP_CLIENT_ID
+ALGORITHMS = ["RS256"]
+
+
+def _parse_der_public_key(b64_der_key: str) -> PublicKeyTypes:
+    """Parses a Base64 encoded DER public key and returns the public key object."""
+    try:
+        der = base64.b64decode(b64_der_key)
+        return serialization.load_der_public_key(der, backend=default_backend())
+    except Exception:
+        logger.exception("Failed to parse DER public key")
+        raise
+
+
+@memory_cache(60 * 30)
+def get_jwks() -> dict:
+    """Retrieves and caches JSON Web Key Sets (JWKS) for token verification."""
+    response = requests.get(JWKS_URL, timeout=5)
+    response.raise_for_status()
+    return response.json()
+
+
+def validate_jwt(token: str, token_type: str) -> dict | None:
+    """Validates the given JWT and returns its claims if valid, otherwise returns None."""
+    token = token.split(" ")[1] if token.startswith("Bearer ") else token
+    # Set required fields based on token type.
+    extra_fields = ["username", "client_id"] if token_type == "access" else ["cognito:username", "email"]  # noqa: S105
+
+    try:
+        return _validate_jwt(token, extra_fields=extra_fields, token_type=token_type)
+    except ExpiredSignatureError:
+        logger.exception("Token has expired", extra={"token_type": token_type})
+    except InvalidTokenError:
+        logger.exception("Invalid token", extra={"token_type": token_type})
+    except Exception:  # pylint: disable=broad-except
+        logger.exception("Error decoding token", extra={"token_type": token_type})
```

Comment:
To aid debugging, should we also consider two more exceptions here:

- **requests.exceptions.HTTPError** which can be raised by line 40
- **json.decoder.JSONDecodeError** which can be raised by line 41

This is just so that the logged information is a bit clearer.